test/FunctionalTests/Infrastructure/MessageHelpers.cs

```diff
@@ -0,0 +1,168 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Google.Protobuf;
+using NUnit.Framework;
+
+namespace Grpc.AspNetCore.FunctionalTests.Infrastructure
+{
+    internal static class MessageHelpers
+    {
+        public const int MessageDelimiterSize = 4;
+
+        public static T AssertReadMessage<T>(byte[] messageData) where T : IMessage, new()
+        {
+            Assert.AreEqual(0, messageData[0]);
+
+            var messageLength = DecodeMessageLength(messageData.AsSpan(1, 4));
+
+            int expectedLength = MessageDelimiterSize + 1 + messageLength;
+
+            Assert.AreEqual(expectedLength, messageData.Length);
+
+            var message = new T();
+            message.MergeFrom(messageData.AsSpan(5).ToArray());
+
+            return message;
+        }
+
+        public static async Task<T> AssertReadMessageAsync<T>(Stream stream) where T : IMessage, new()
+        {
+            var messageData = await ReadMessageAsync(stream);
+            if (messageData == null)
+            {
+                return default;
+            }
+
+            var message = new T();
+            message.MergeFrom(messageData);
+
+            return message;
+        }
+
+        public static async Task<byte[]> ReadMessageAsync(Stream stream)
```

Comment:
I think most of these points are addressed by https://github.com/aspnet/grpc-dotnet/pull/16. Could add more tests though.

Reply:
I've added some sanity check tests that the raw bytes are correct - https://github.com/aspnet/grpc-dotnet/pull/16/commits/416d64889934e969fade2777b45665f6aa68e356

As part of the pipelines PR I'll clean up the test helpers to use the core app logic instead of copy pasted code